### PR TITLE
Handle most openssl3 deprecations

### PR DESCRIPTION
**Pull request by frozencemetery**
Right now krb5 claims to support all the way back to openssl 1.0.0.  Though I haven't tested that far back, I have tried to preserve compatibility.  That said, 1.0.0 was 11 years ago; it might be nice to raise the minimum version.

Caveats to this PR:

- Currently this needs a version from git, as Camellia CTS support has not yet landed in a beta.

- This doesn't really address PKINIT, where the DH_* are deprecated.  It seems possible albeit involved to port - openssl3 does have the ability to set OSSL_PKEY_PARAM_FFC_P, OSSL_PKEY_PARAM_FFC_G, and OSSL_PKEY_PARAM_FFC_Q.

- Deprecation of DES_set_odd_parity() is left unaddressed.  We're on a path to remove it from krb5 anyhow, I patch out DES3 downstream, and OpenSSL was not interested in adding a non-deprecated equivalent or removing the deprecation.

- This supersedes #979, as it was difficult to test locally without the provider glue code.  I've also switched to just checking OPENSSL_VERSION_NUMBER rather than for specific functions, and squashed it into a single commit.

### CODE REVIEW

**Comment by satmandu** [comment]
Now that OpenSSL 3.0.0 is out, it would be nice to see this or an equivalent merged...

**Comment by satmandu** [comment]
Looks like krb5 from git is compiling w/ OpenSSL 3.0.0 w/o any additional patches.

**Comment by greghudson** [comment]
Notes on the first commit:

* What happens if we just don't call OSSL_PROVIDER_unload()?  Is that a memory leak, or will libcrypto deinitialization reclaim that memory?  I've taken a look at the libcrypto code but haven't found the answer yet.

* OSSL_PROVIDER_load()'s implementation claims to disable auto-loading of defaults.  Could that be a problem for other users of libcrypto?

* If libk5crypto is explicitly unloaded before the process exits, could we cause a crash at process exit trying to call the no-longer-valid function?  It appears that on most (all?) Unix platforms, atexit() handlers are called and unregistered on library unload, so no.  But apparently not on Windows; OpenSSL takes care to use _onexit() rather than atexit() for this reason.

**Comment by greghudson** [comment]
I see a memory leak in this test program:

    lh = dlopen("libcrypto.so.3", RTLD_NOW);
    loadfn = dlsym(lh, "OSSL_PROVIDER_load");
    (*loadfn)(NULL, "legacy");
    dlclose(lh);

So I guess we do need the unload step to avoid a memory leak on repeated loads and unloads of libk5crypto.  I'm a little nervous about the mechanism.

I assume that you tried using a library finalizer and empirically ran into the ordering issue that OpenSSL had already run its atexit() cleanup handler, on Linux.  I'm a little unclear about why that is; I guess even though atexit() handlers run at library unload time, they all still run before library finalizers at process exit.

Separately, I want to look for other places than cbc_enc()/cbc_enc() to ensure that the library initializer runs.  Usually we put those calls closer to the API surface.

**Comment by frozencemetery** [comment]
> I assume that you tried using a library finalizer and empirically ran
> into the ordering issue that OpenSSL had already run its atexit()
> cleanup handler, on Linux.  I'm a little unclear about why that is; I
> guess even though atexit() handlers run at library unload time, they
> all still run before library finalizers at process exit.

Yes, that's what happened.

**Comment by greghudson** [comment]
The second commit puts OpenSSL-using code into crypto/krb conditionally, damaging the organizational principles of libk5crypto.

Simply defining crypto_int.h interfaces for key derivation is insufficient to address this concern, as crypto back-ends are currently all-or-nothing with no way to selectively borrow from the code in `builtin`.  So we can't currently make the openssl module use the built-in KDFs for OpenSSL <3 and the OpenSSL ones for >=3.

I'm preparing to do some build system work to address this problem.  My current thinking is that we make the build system descend into builtin and openssl always, and conditionalize each file in both directories on some define (like BUILTIN_KDF or OPENSSL_CAMELLIA).  Then crypto_int.h decides which defines to enable based on the crypto back-end and OpenSSL version.  This would also address some ancillary problems: `openssl` would no longer be required to match the directory structure of `builtin`, and "make depend" would generate all of the libk5crypto dependencies regardless of build configuration.

I'm also thinking of discarding the Fortuna PRNG and associated logic, for simplicity.  With AES_encrypt() etc. deprecated in OpenSSL 3, there are no longer failure-free interfaces to AES256 and SHA256 in OpenSSL, which the code currently depends on.  Alternatively, we could make using the openssl crypto back-end imply using the os PRNG back-end.

**Comment by frozencemetery** [comment]
I had no issues shipping os without fortuna, though [Debian still enables it](https://salsa.debian.org/debian/krb5/-/blob/master/debian/rules#L42) - I seem to remember this being an intentional choice on their part, though those concerns may have abated.

One other option to consider might be OpenSSL's OSSL_rand().  I wasn't personally keen on it, but it might assuage concerns about not having an in-process PRNG, if there are any.

**Comment by greghudson** [comment]
If I'm successful at reworking the build system to allow selective crypto back-end overrides, an alternative to the first commit would be to use the built-in MD4 and RC4 for OpenSSL 3 instead of enabling the legacy provider.

**Comment by kaduk** [comment]
You definitely want your own libctx -- this type of scenario is a primary motivator for creating the openssl libctx.

Manually loading a provider for a libctx disables the silent fallback where the default provider is loaded if there are no other providers available, which is very disruptive if you load into the default (`NULL`) libctx.

My understanding (without checking) is that Greg's observation of a memory leak in the absence of `OSSL_PROVIDER_unload()` is correct.

I think (but again, didn't check) that openssl went with atexit() because openssl is used on so many platforms that trying to get the non-portable initializers/finalizers working would be too hard to support.  But yes, it has problems

**Comment by greghudson** [comment]
Is there a way to use a specific library context with EVP_EncryptInit/Update/Final, other than OSSL_LIB_CTX_set0_default()?

(At the moment I favor the approach of using the builtin RC4 and MD4 implementations when compiled against OpenSSL 3, which renders this question moot.  But that might change if I see pushback from downstreams who build against OpenSSL.)

File: src/lib/crypto/openssl/cmac.c [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+                      krb5_data *output)
+{
+    int ok;
+    EVP_MAC *mac = NULL;
+    EVP_MAC_CTX *ctx = NULL;
+    OSSL_PARAM params[2];
+    size_t i = 0, md_len;
+    char *cipher;
+
+    if (enc->block_size != BLOCK_SIZE)
```
[the remark this answers is not in the document]
Comment (greghudson) [reply src/lib/crypto/openssl/cmac.c:53]: I must have put this in without re-checking the OpenSSL 3 build.  I'll leave this check out.  The subsequent cipher translation precludes using the function with an inappropriate cipher.

**Pull request merged**